### PR TITLE
Fixed bug with emojis and action tag reaction due to incorrect character movement across emojis (Resolves #1882)

### LINKED ISSUE #1882: Emoji at beginning of paragraphs cause problems with Tags
**Package Version**
1. super_editor `main` channel
2. Flutter `main` latest

**To Reproduce**
1. Launch the example app on macos
3. Go to the Action Tags demo
4. Try to insert the 🔴 emoji
5. See the error

**Actual behavior**
No emoji is inserted. Error thrown (see below)

**Expected behavior**
The emoji is inserted correctly.

**Platform**
macOS

**Flutter version**
```
[!] Flutter (Channel main, 3.20.0-17.0.pre.99, on macOS 14.2.1 23C71 darwin-arm64, locale en-US)
Dart SDK version: 3.4.0-190.0.dev (dev) (Thu Feb 29 08:14:33 2024 -0800) on "macos_arm64"
```

**Error**
```
======== Exception caught by services library ======================================================
The following _TypeError was thrown during method call TextInputClient.updateEditingStateWithDeltas:
Null check operator used on a null value

When the exception was thrown, this was the stack: 
#0      TagFinder.findTagAroundPosition (package:super_editor/src/default_editor/text_tokenizing/tags.dart:40:95)
#1      ActionTagComposingReaction.react (package:super_editor/src/default_editor/text_tokenizing/action_tags.dart:299:37)
#2      Editor._reactToChanges (package:super_editor/src/core/editor.dart:218:16)
#3      Editor.execute (package:super_editor/src/core/editor.dart:161:9)
#4      TextDeltasDocumentEditor._insertPlainText (package:super_editor/src/default_editor/document_ime/document_delta_editing.dart:342:12)
#5      TextDeltasDocumentEditor.insert (package:super_editor/src/default_editor/document_ime/document_delta_editing.dart:294:23)
#6      TextDeltasDocumentEditor._applyInsertion (package:super_editor/src/default_editor/document_ime/document_delta_editing.dart:171:5)
#7      TextDeltasDocumentEditor.applyDeltas (package:super_editor/src/default_editor/document_ime/document_delta_editing.dart:80:9)
#8      DocumentImeInputClient.updateEditingValueWithDeltas (package:super_editor/src/default_editor/document_ime/document_ime_communication.dart:219:30)
#9      DeltaTextInputClientDecorator.updateEditingValueWithDeltas (package:super_editor/src/default_editor/document_ime/ime_decoration.dart:129:14)
#10     TextInput._handleTextInputInvocation (package:flutter/src/services/text_input.dart:1892:63)
#11     TextInput._loudlyHandleTextInputInvocation (package:flutter/src/services/text_input.dart:1781:20)
#12     MethodChannel._handleAsMethodCall (package:flutter/src/services/platform_channel.dart:571:55)
#13     MethodChannel.setMethodCallHandler.<anonymous closure> (package:flutter/src/services/platform_channel.dart:564:34)
#14     _DefaultBinaryMessenger.setMessageHandler.<anonymous closure> (package:flutter/src/services/binding.dart:581:35)
#15     _invoke2 (dart:ui/hooks.dart:344:13)
#16     _ChannelCallbackRecord.invoke (dart:ui/channel_buffers.dart:45:5)
#17     _Channel.push (dart:ui/channel_buffers.dart:135:31)
#18     ChannelBuffers.push (dart:ui/channel_buffers.dart:343:17)
#19     PlatformDispatcher._dispatchPlatformMessage (dart:ui/platform_dispatcher.dart:750:22)
#20     _dispatchPlatformMessage (dart:ui/hooks.dart:257:31)
call: MethodCall(TextInputClient.updateEditingStateWithDeltas, [28, {deltas: [{selectionBase: 4, oldText: . , selectionAffinity: TextAffinity.downstream, deltaEnd: 2, deltaText: 🔴, deltaStart: 2, composingExtent: -1, selectionIsDirectional: false, selectionExtent: 4, composingBase: -1}]}])
====================================================================================================
```

**Additional context**
Using tags, we are also seeing this in Superlist when you:

  1. Insert an emoji at the beginning of a paragraph
  2. Press left arrow key (to move the arrow before the emoji)
  3. The same `findTagAroundPosition` throws an error

### PR DESCRIPTION
**Pull request by matthew-carroll**
Fixed bug with emojis and action tag reaction due to incorrect character movement across emojis (Resolves #1882)

The issue was that the action tag reaction was assuming, in a couple places, that every character only occupies a single codepoint.

### CODE REVIEW

**Comment by BazinC** [comment]
Hey @matthew-carroll , your modification in TagFinder won't solve the whole issue with emojis. 
See video below captured on your branch.

https://github.com/superlistapp/super_editor/assets/6229343/37ebc067-c81a-4e1e-bb56-e85a8e3fb7fe

- Action Tag example
- Type emoji '💙' -> ok
- Move the cursor back one character, before the emoji
- Type the action trigger character '/'
- TagFinder leads to not well-formed UTF-16

Can you please consider reviewing my [PR](https://github.com/superlistapp/super_editor/pull/1879) here, which solves both issues? Thank you :)

File: super_editor/lib/src/default_editor/text_tokenizing/tags.dart [hunk 1]
```diff
@@ -56,15 +61,27 @@ class TagFinder {
...
-      final downstreamCharacter = rawText[downstreamCharacterIndex];
+    // Push downstream until we reach the end of the token, or the end of the available text.
+    //
+    // WARNING: This text offset push needs to move by character, which might be a multi-code-point
+    // character, like an emoji.
+    int? nextCharacterOffset = rawText.moveOffsetDownstreamByCharacter(tokenEndOffset);
+    while (nextCharacterOffset != null && nextCharacterOffset < rawText.length) {
+      if (nextCharacterOffset >= rawText.length) {
+        break;
+      }
```
Comment (angelosilvestre) [inline super_editor/lib/src/default_editor/text_tokenizing/tags.dart:72]: Is this if statement necessary? We already check this in the `while` condition.

File: super_editor/test/super_editor/text_entry/tagging/action_tags_test.dart [hunk 2]
```diff
@@ -650,6 +650,56 @@ void main() {
         );
       });
     });
+
+    group("emojis >", () {
+      // Based on bug https://github.com/superlistapp/super_editor/issues/1882
+      testWidgetsOnArbitraryDesktop("can appear at the beginning of text", (tester) async {
```
Comment (angelosilvestre) [inline super_editor/test/super_editor/text_entry/tagging/action_tags_test.dart:656]: Does this test belongs to this file? It seems we don't do anything related to tags in this test.

**Review by angelosilvestre — COMMENTED** [review]
We should probably take a look at @BazinC 's PR.

**Comment by matthew-carroll** [comment]
I'm gonna close this in favor of https://github.com/superlistapp/super_editor/pull/1879 where we'll attempt to review and merge a different fix for the same problem.

**Pull request closed without merging**